src/Service/NixGenerator.php

```diff
@@ -0,0 +1,392 @@
+                            sprintf(
+                                "Package '%s' has dist-type '%s' which is not" .
+                                ' supported by the Nixify plugin',
+                                $package->getPrettyName(),
+                                $package->getDistType()
+                            )
+                        );
+
+                    break;
+            }
+        }
+    }
+
+    /**
+     * Generates Nix files based on the lockfile and cache.
+     */
+    public function generate(array $collected): void
+    {
+        $package = $this->composer->getPackage();
+
+        // Build Nix code for cache entries.
+        $cacheEntries = array_filter(
+            $collected,
+            static fn (array $info): bool => 'cache' === $info['type']
+        );
+
+        // Build Nix code for local entries.
+        $localEntries = array_filter(
+            $collected,
+            static fn (array $info): bool => 'local' === $info['type']
+        );
+
+        // If the user bundled Composer, use that in the Nix build as well.
+        $cwd = getcwd() . '/';
+        $composerPath = realpath($_SERVER['PHP_SELF']);
+
+        $composerPath = substr($composerPath, 0, strlen($cwd)) === $cwd
+            ? substr($composerPath, strlen($cwd))
+            : null;
+
+        $jsonData = [
+            'cacheEntries' => $cacheEntries,
+            'composerPath' => $composerPath,
+            'localEntries' => $localEntries,
+            'projectName' => $package->getExtra()['nix-expr-path'] ?? 'composer-project.nix',
+        ];
+
+        $searchNreplace = [
+            '{{json}}' => json_encode(
+                array_filter($jsonData),
+                JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES
+            ),
+        ];
+
+        file_put_contents(
+            $package->getExtra()['nix-expr-path'] ?? 'composer-project.nix',
+            str_replace(
+                array_keys($searchNreplace),
+                array_values($searchNreplace),
+                file_get_contents(__DIR__ . '/../res/composer-project.nix')
+            )
+        );
+
+        // Generate default.nix if it does not exist yet.
+        $generateDefaultNix = $package->getExtra()['generate-default-nix'] ?? true;
+
+        if ($generateDefaultNix && !file_exists('default.nix') && !file_exists('flake.nix')) {
+            file_put_contents('default.nix', file_get_contents(__DIR__ . '/../res/default.nix'));
+            $this->io->writeError(
+                '<info>A minimal default.nix was created. You may want to customize it.</info>'
+            );
+        }
+    }
+
+    /**
+     * Preload packages into the Nix store.
+     */
+    public function preload(array $collected): void
+    {
+        $tempDir = sprintf(
+            '%s.nixify-tmp-%s',
+            $this->cacheDir,
+            substr(md5(uniqid('', true)), 0, 8)
+        );
+
+        $toPreload = array_filter(
+            array_map(
```

Comment:
I don't think these nested calls improve readability.

I think this idea can work if you use functional style helpers, like Laravel's collection helpers, but it's not worth the dependency. IMO, intermediate variables or just the old for-loop is a lot more readable.